srvutil/metrics.go

```diff
@@ -56,41 +80,56 @@ func (w *hijackableRecorder) Hijack() (net.Conn, *bufio.ReadWriter, error) {
 	return w.httpRecorder.ResponseWriter.(http.Hijacker).Hijack()
 }
 
-func newHTTPRecorder(w http.ResponseWriter) loggableHTTPRecorder {
-	recorder := httpRecorder{ResponseWriter: w}
+func newHTTPRecorder(w http.ResponseWriter, bodyLogPredicate BodyLogPredicateFunc) loggableHTTPRecorder {
+	recorder := httpRecorder{ResponseWriter: w, bodyLogPredicate: bodyLogPredicate}
 	if _, ok := w.(http.Hijacker); ok {
 		return &hijackableRecorder{recorder}
 	}
 	return &recorder
 }
 
-// RequestMetricsMiddleware records the time taken to serve a request.
+type RequestMetricsMiddlewareConfig struct {
+	BodyLogPredicate BodyLogPredicateFunc
+}
+
+// NewRequestMetricsMiddleware records the time taken to serve a request, and logs request and response data.
 // Example tags: statusClass:2xx, statusCode:200
 // Should be added as a middleware after RequestContextMiddleware to benefit from its tags
-func RequestMetricsMiddleware(next http.Handler) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		ctx := r.Context()
+func NewRequestMetricsMiddleware(c RequestMetricsMiddlewareConfig) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx := r.Context()
 
-		recorder := newHTTPRecorder(w)
-		ctx = statsd.WatchingTagLoggable(ctx, recorder)
-		r = r.WithContext(ctx)
+			recorder := newHTTPRecorder(w, c.BodyLogPredicate)
+			ctx = statsd.WatchingTagLoggable(ctx, recorder)
+			r = r.WithContext(ctx)
 
-		reqHeaders := redact.Headers(r.Header)
+			reqHeaders := redact.Headers(r.Header)
 
-		log(ctx, nil).
-			WithField("method", r.Method).
-			WithField("headers", reqHeaders).
-			Info("http request")
+			log(ctx).
+				WithField("method", r.Method).
+				WithField("headers", reqHeaders).
+				Info("http request")
 
-		metrics.HTTPRequest.Time(ctx, func() error {
-			next.ServeHTTP(recorder, r)
-			return nil
-		})
+			metrics.HTTPRequest.Time(ctx, func() error {
+				next.ServeHTTP(recorder, r)
+				return nil
+			})
+
+			resHeaders := redact.Headers(w.Header())
+
+			logger := log(ctx).
+				WithField("headers", resHeaders)
 
-		resHeaders := redact.Headers(w.Header())
+			if body := recorder.ResponseBody(); body != nil {
+				logger = logger.WithField("responseBody", *body)
+			}
 
-		log(ctx, nil).
-			WithField("headers", resHeaders).
-			Info("http response")
-	})
+			logger.Info("http response")
+
+		})
+	}
 }
+
+// RequestMetricsMiddleware is here for backwards compatibility.
+var RequestMetricsMiddleware = NewRequestMetricsMiddleware(RequestMetricsMiddlewareConfig{})
```

Comment:
Because structs are passed by value in Go, they get copied/duplicated every time they are passed as an argument.
Also, only references to structs can satisfy interfaces. 

For this reason, it's customary to pass struct pointers, not values.

Reply:
I changed to use pointers, please have another look.

> Also, only references to structs can satisfy interfaces.

What do you mean by this? The following is a non-reference struct that satisfies an interface (`error`):

```go
type MyErr struct {}

func (m MyErr) Error() string {
  return "my error"
}
```

See https://play.golang.org/p/0i_e3ZaX-H9.

Reply:
Ah! Interesting! My mistake. That's because here you're using `(m MyErr)` as a receiver. I would normally be using `(m* MyErr)`, also to creating copies of the structs for the reason above. That's why I was getting interface mismatch all this time 😄 

In any case, it's still customary to use pointers for structs. You can find plenty of articles like https://flaviocopes.com/golang-methods-receivers/ suggesting the same

Reply:
Yep, I agree, pointers here is nicer.